src/schema/v2/me/user_flags.ts

```diff
@@ -0,0 +1,94 @@
+import {
+  GraphQLError,
+  GraphQLFieldConfig,
+  GraphQLScalarType,
+  Kind,
+  ObjectValueNode,
+  print,
+  ValueNode,
+} from "graphql"
+import Maybe from "graphql/tsutils/Maybe"
+import { camelCase } from "lodash"
+import { ResolverContext } from "types/graphql"
+
+export const UserFlagsType = new GraphQLScalarType({
+  name: "UserFlagsType",
+  description:
+    "The user's user_flags. Values of keys in user_flags can be anything and keys may not be known beforehand.",
+  serialize: (value) => value,
+  parseValue: (value) => value,
+  parseLiteral: (ast, variables) => {
+    if (ast.kind !== Kind.OBJECT) {
+      throw new TypeError(
+        `Object cannot represent non-object value: ${print(ast)}`
+      )
+    }
+
+    return parseObject("Object", ast, variables)
+  },
+})
+
+export const userFlagsResolver = async (root, _options, loaders) => {
+  const { meLoader } = loaders
+  const { user_flags: userFlags } = root
+  if (!meLoader) {
+    throw new Error("You need to be signed in to perform this action")
+  }
+
+  const res = {}
+  const keys = Object.keys(userFlags)
+  keys.forEach((key) => {
+    res[camelCase(key)] = userFlags[key]
+  })
+  return res
+}
+
+export const UserFlags: GraphQLFieldConfig<any, ResolverContext> = {
+  type: UserFlagsType,
+  description:
+    "Hash of arbitrary keys/values. Useful for transient or UI-related preferences.",
+  resolve: async (root, options, loaders) => {
+    return await userFlagsResolver(root, options, loaders)
+  },
+}
+
+// custom scalar helpers for UserFlags
+const parseObject = (
+  typeName: string,
+  ast: ObjectValueNode,
+  variables: Maybe<{ [key: string]: any }>
+) => {
+  const value = {}
+  ast.fields.forEach((field) => {
+    value[field.name.value] = parseLiteral(typeName, field.value, variables)
+  })
+
+  return value
+}
+
+const parseLiteral = (
+  typeName: string,
+  ast: ValueNode,
+  variables: Maybe<{ [key: string]: any }>
+) => {
+  switch (ast.kind) {
```

Comment:
I wonder if it makes things easier or more difficult to deserialize values at this point or just return them as a string 🤔 

Wouldn't we need to assert the type again when reading it in a TypeScript API client in most cases?

For example: 
```
const myNumberFlag = graphqlResponse.userFlags.myNumberFlag
// type of myFlag is now: string | boolean | number | object | null .....

const isPositiveNumber = (myNumberFlag as number) > 0
```

@MounirDhahri I would be interested to hear what you think about this?

Reply:
I am not sure I understand. The userFlag type is a custom scalar one. When doing the parseLiteral for this type I am just assuming that there might be a possibility of a key having an entire subset of object, hence necessitating this recursion: https://github.com/artsy/metaphysics/pull/4192/files#diff-79baba1fc0b350dbe7117a48d79bb8656f9501f41acecfe07dbe5ab8b31f9b83R82 .
I think it is inevitable that if a client expects a number, then they have to reassert the type, because we cannot type such a dynamic object as userFlags. Hope I'm correctly interpreting the concern.

Reply:
@olerichter00 that's a good point. Ideally, we wouldn't be in a position to return a value that can be of different types. I am wondering if we actually need to also support the other types instead of returning a string or null. My main thinking behind that is that it would be easier for the client to be always expecting a string for the user_flag instead of a scalar and we would be basing front end logic on the string value (or its non existence) instead of the returned value type + its value (or its non existence).

Reply:
Ohhh, I see. I get it now.
But the custom scalar relates to the entire userFlags. In the schema this should appear as `scalar MyType`, it wouldn't know what the internals are and so there is no benefit to the client so forcing a string would not benefit the client, I think. Right?

Reply:
I'm in favor of just returning a string and then deserializing in the client (example 2) to avoid type assertion (see example 1). It would also make the logic in MP easier. 

And I'd love to hear more opinions from other engineers :)

Reply:
Since [the endpoint that sets user flags forces values into strings](https://github.com/artsy/gravity/blob/92896cef4992601d7a1298c2028f0987f683987c/app/api/v1/me_endpoint.rb#L128), I think the most consistent treatment in Metaphysics is to return those strings. Clients can then [cautiously] convert the values based on the particular keys they might have knowledge about. A[ test confirms](https://github.com/artsy/gravity/blob/92896cef4992601d7a1298c2028f0987f683987c/spec/api/v1/me_endpoint_spec.rb#L568-L571) that, even if the value is supplied as an integer, it's returned as a string.

Note that these strings could obviously be abused to store serializations of complex objects, but I wouldn't recommend that. Numbers seem OK though.

Reply:
I think we can trust that userFlags set via the endpoints will always be a string. But I worry that if any service in the BE sets a user flag that is a complex object or not a string, then it will be assumptive of Metaphysics to force a type of string on the values, because some values may **_not_** be a string. And that is why I favour the approach of the custom scalar with the recursive parseLiteral and clients having to reassert if need be - though this would be rare, because clients have a foreknowledge what kind of value the key they want to check would hold.